PP-747: Modify the sched object to have additional options to configure pbs_sched

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-747](https://pbspro.atlassian.net/browse/PP-747)**

#### Problem description
This ticket is one of the major user-stories of the following EPIC.
**[PP-337](https://pbspro.atlassian.net/browse/PP-337)** Multiple schedulers servicing the PBS cluster

The current user-story actually implements major interface for the MultiSched feature. After this we can actually run multiple instances of scheduler each serving a set of partitions in parallel which actually accomplishes the following goal.

As clusters get larger and workloads vary it is becoming critical that the jobs get evaluated in as short as time possible to ensure that the correct workload is being run. Using multiple schedulers to address this issue can allow for different scheduling policies and quicker turnaround time for large number of jobs or nodes.

#### Solution description
Nodes and Queues in PBS now have a special attribute called partition. This can be useful in combining or grouping a set of nodes and queues together.
For example we can set partition attribute on nodes and queues as follows.

qmgr -c "set node n1 partition=p1"
qmgr -c "set node n2 partition=p1"
qmgr -c "set queue q1 partition = p1"
Here we grouped n1,n2 and q1 under partition p1. Similarly we can come up with multiple partitions as per our requirements. Let us say we have one more partition p2 which groups nodes n3,n4,n5 and q2.

A scheduler other than the default scheduler can be associated to one or more partitions as given in the following example.
qmgr -c "set sched sc1 partition=p1"
To associate more than one partition,
qmgr -c "set sched sc1 partition += p2"

Now the scheduler sc1 can serve two partitions p1 and p2.

An important note here is we should configure the scheduler first in the server then only it can be started. If a server does not know a scheduler id then server rejects it and that scheduler won't come up.

If we try to start a scheduler which is not associated with at least one partition then the following error is thrown while it is coming up.
"Scheduler should contain at least single partition"
Scheduler will only come up only if it is fully configured at the server.

Once the scheduler comes up successfully it is ready to serve partitions p1, p2 in this case. All the scheduler features should work as expected. Some of them are as follows

We can submit jobs to q1 which are handled by sc1 scheduler in this case.
Example: qsub -q q1 -- /bin/sleep 100
This job should be served by sc1 and we can observe the same in the log file in sched_logs_sc1
directory. We can also observe the same in pbsnodes -av output i.e. this job should only go to one of
the nodes that is part of the partition p1.
qrun of a job submitted to the above queue or partition should work.
suspend/resume functionality should work.
Array jobs can also be submitted and its functionality should work.
One can create resources of various types and each scheduler should work as expected and schedule
using these resources.
Reservations until we implement next interface will now only be served by default scheduler.
Have done all the code changes in such a way that backward compatibility with the default scheduler is
maintained.
server's scheduler_iteration and scheduling attributes are still kept even though we have these
attributes for the scheduler object also. This is done to ensure the backward compatibility. If anybody
sets these attributes at one object then they are automatically reflected in another object.
Certain operations like changing the scheduler port and sched_priv directory etc are not allowed on
default scheduler to maintain the backward compatibility.
Dynamically one can associate multiple partitions to the scheduler. We no need to restart or SIGHUP
the corresponding scheduler.
Dynamically one can dissociate a partition from the scheduler.
One can change sched_priv/sched_log directory dynamically without restarting or SIGHUPing the
scheduler.
If a partition is associated with one scheduler then it can not be associated to any other scheduler.
Scheduler at any point in time if it fails in accepting the new configuration then it switches back to its
old configuration.
All operations for queues that are having no partitions should be served by default scheduler.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

suresh-thelkar commented:
Had to raise this pull request since we closed the following pull request because GitHub seems does not accept if the size of pull request is reaching some threshold.
https://github.com/PBSPro/pbspro/pull/454

Please refer back and forth between the above pull request and the current one to get all the details.

varunsonkar commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.manager(MGR_CMD_SET, QUEUE, p3, id='wq3')
+        p4 = {'partition': 'P4'}
+        self.server.manager(MGR_CMD_SET, QUEUE, p4, id='wq4')
+        a = {'resources_available.ncpus': 2}
```
line: 88-102
Please validate with "expect=True" for all MGR_CMD_CREATE/MGR_CMD_SET commands

varunsonkar commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.manager(MGR_CMD_SET, NODE, p2, id='vnode[1]', expect=True)
+        self.server.manager(MGR_CMD_SET, NODE, p3, id='vnode[2]', expect=True)
+        self.server.manager(MGR_CMD_SET, NODE, p4, id='vnode[3]', expect=True)
+
```
When I see your test setUp it looks very heavy with lot of multischeds created for all test cases.It should be lighter. Have stuff only which is actually common.
For rest create a common setup function which could be called as per the requirement in the test cases.

varunsonkar commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.manager(MGR_CMD_SET, NODE, p4, id='vnode[3]', expect=True)
+
+    def sched_configure(self, sched_name, sched_port=None, sched_home=None):
+        pbs_home = self.server.pbs_conf['PBS_HOME']
```
For functions please add the details about it and the attributes in the docstring as per the PTL guidelines

varunsonkar commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+                             os.path.join(sched_home, sched_logs_dir),
+                             recursive=True)
+        self.server.schedulers[sched_name].start(sched_port, sched_home)
+
```
function "sched_configure" you are writing two times in functional test cases and interface test cases.
Suggest to put in the common place "PBSTestSuite" and may be make it more generic.

varunsonkar commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -156,9 +182,129 @@ def test_set_and_unset_sched_attributes(self):
+        self.server.manager(MGR_CMD_LIST, SCHED, id='default')
+        self.server.expect(SCHED, {'scheduler_iteration': 500},
+                           id='default', max_attempts=10)
+
```
In the test suite the negative scenarios for the set/unset/create new attributes of multi-sched are missing.
Please add them.

suresh-thelkar commented:
@bhroam / @arungrover 
This is just to update you that this is the new pull request for PP-747 and I have also given the old PR link in this PR for reference. 
In order to raise this new pull request and make sure that Travis/appveryor runs fine I had to re-base my branch with the master latest. In the process of resolving merge conflicts and fixing all warnings which are now treated as errors I have added couple of commits today. Please look into these latest commits also.

bhroam commented on src/scheduler/data_types.h:
```diff
@@ -602,6 +603,7 @@ struct node_info
 	node_info *svr_node;		/* ptr to svr's node if we're a resv node */
 	node_partition *hostset;      /* other vnodes on on the same host */
 	node_scratch nscr;            /* scratch space local to node search code */
+	char *partition;	      /*  partition to which node belongs to */
```
It looks like there are two spaces after the /*

suresh-thelkar replied:
Done

bhroam commented on src/scheduler/globals.h:
```diff
@@ -41,7 +41,7 @@ extern "C" {
 
 #include "data_types.h"
-
+#include "limits.h"
```
Why is this needed in globals.h?

suresh-thelkar replied:
This header is needed to get the definition of _POSIX_PATH_MAX

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -208,8 +209,10 @@ query_nodes(int pbs_sd, server_info *sinfo)
 	}
 #endif /* localmod 049 */
 
+	my_partitions = break_comma_list(partitions);
```
break_comma_list() is allocating memory, you might want to check for NULL

suresh-thelkar replied:
Done

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -208,8 +209,10 @@ query_nodes(int pbs_sd, server_info *sinfo)
+
 	cur_node = nodes;
-	for (i = 0; cur_node != NULL; i++) {
+	for (i = 0, nidx=0; cur_node != NULL; i++) {
```
Add spaces around the =

suresh-thelkar replied:
Done

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -222,6 +225,32 @@ query_nodes(int pbs_sd, server_info *sinfo)
+				}
+
+			}
+		}
```
The scheduler style avoids to use of a continue if at all possible.  It is definitely possible here.  Move this code into a function called something like node_in_partition().  It can return 1 if the node is in the partition.  You can easily use that in an if().  Plus, you're leaking memory.  If you don't put ninfo in ninfo_arr, it will never be freed.

suresh-thelkar replied:
Implemented. Also done similar changes for querying/filtering queues info.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -310,6 +339,13 @@ query_node_info(struct batch_status *node, server_info *sinfo)
+		else if(!strcmp(attrp->name, ATTR_partition)) {
+			ninfo->partition = string_dup(attrp->value);
+			if (ninfo->partition == NULL) {
+				log_err(errno, "node_queue_info", MEM_ERR_MSG);
```
don't use log_err(), use schdlog().  Plus use \_\_func\_\_

suresh-thelkar replied:
Have changed to use \__func__. But since it is system error I think it is better to use log_err() with errno. Any specific reason why only schdlog() is to be used in this case ?

bhroam replied:
The scheduler uses schdlog().  If you don't use it, then log_filter will fail.  Please use schdlog().  Every malloc() failure in the scheduler uses the same message (MEM_ERR_MSG) with schdlog().

suresh-thelkar replied:
Thanks for letting me know this. I have made the changes accordingly. It seems there are several places where still log_err() is used with MEM_ERR_MSG which lead to my initial confusion.

bhroam replied:
There are?  There shouldn't be.  I'm sorry for the confusion.  Everything in the scheduler should use schdlog() until we move to a more uniform logging framework for all of PBS (or even move the scheduler to log_events).

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -596,6 +633,10 @@ free_node_info(node_info *ninfo)
 
+		if (ninfo->partition != NULL) {
+			free(ninfo->partition);
+		}
```
Remove the {}

suresh-thelkar replied:
Done

bhroam commented on src/scheduler/queue_info.c:
```diff
@@ -173,9 +175,11 @@ query_queues(status *policy, int pbs_sd, server_info *sinfo)
 	}
 	qinfo_arr[0] = NULL;
 
+	my_partitions = break_comma_list(partitions);
```
Like above, you probably want to check for NULL

suresh-thelkar replied:
Done

bhroam commented on src/scheduler/queue_info.c:
```diff
@@ -184,6 +188,32 @@ query_queues(status *policy, int pbs_sd, server_info *sinfo)
+
+			}
+		}
+
```
Like my above comment, put this code in a function and avoid the use of continues

suresh-thelkar replied:
Implemented.

bhroam commented on src/scheduler/queue_info.c:
```diff
@@ -381,6 +411,13 @@ query_queue_info(status *policy, struct batch_status *queue, server_info *sinfo)
+		else if(!strcmp(attrp->name, ATTR_partition)) {
+			qinfo->partition = string_dup(attrp->value);
+			if (qinfo->partition == NULL) {
+				log_err(errno, "query_queue_info", MEM_ERR_MSG);
```
Use schdlog() and \_\_func\_\_

suresh-thelkar replied:
Have changed to use \__func__. And please look into my above response for the rest of the comment.

suresh-thelkar replied:
Thanks for letting me know this. I have made the changes accordingly. It seems there are several places where still log_err() is used with MEM_ERR_MSG which lead to my initial confusion.

bhroam commented on src/scheduler/queue_info.c:
```diff
@@ -825,6 +862,9 @@ free_queue_info(queue_info *qinfo)
 	}
+	if (qinfo->partition != NULL) {
+		free(qinfo->partition);
+	}
```
remove the {}

suresh-thelkar replied:
Done

bhroam commented on src/server/req_stat.c:
```diff
@@ -445,77 +442,23 @@ req_stat_que(struct batch_request *preq)
+			rc = status_que(pque, preq, &preply->brp_un.brp_status);
+			if (rc != 0) {
+				if (rc == PBSE_PERM)
+					rc = 0;
```
Why do you check for rc == PBSE_PERM here, but not on the other side of the if()?

suresh-thelkar replied:
Now that we have moved the filtering of queues/nodes to scheduler, I did not touch anything in this part of the code.

bhroam commented on src/server/req_stat.c:
```diff
@@ -638,77 +578,18 @@ req_stat_node(struct batch_request *preq)
-	} else {
-		/*get status of all nodes     */
-		for (i = 0; i < svr_totnodes; i++) {
+		for (i=0; i<svr_totnodes; i++) {
```
put spaces around the = and <

suresh-thelkar replied:
Done

bhroam commented on src/server/req_stat.c:
```diff
@@ -638,77 +578,18 @@ req_stat_node(struct batch_request *preq)
-			return;
-		}
-	}
+	} else {			/*get status of all nodes     */
```
put a space after the /* and remove the extra spaces before the */

suresh-thelkar replied:
Done

bhroam commented on src/server/req_stat.c:
```diff
@@ -638,77 +578,18 @@ req_stat_node(struct batch_request *preq)
 	CLEAR_HEAD(preply->brp_un.brp_status);
 
-	memset(&attr, 0, sizeof(attr));
+	if (type == 0) {		/*get status of the named node*/
```
put spaces around the comment text

suresh-thelkar replied:
Implemented.

bhroam commented on src/scheduler/fifo.c:
```diff
@@ -2355,6 +2351,46 @@ next_job(status *policy, server_info *sinfo, int flag)
+	if (*dest != NULL)
+		free(*dest);
+
+	if (src !=NULL) {
```
Put a space after =

suresh-thelkar replied:
Done

bhroam replied:
Is this coming in a future commit?  I don't see it done.

suresh-thelkar replied:
yes

bhroam commented on src/scheduler/fifo.c:
```diff
@@ -2355,6 +2351,46 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ */
+static int
+copy_attr_value(char **dest, char *src)
```
This function should be in misc.c, not fifo.c

suresh-thelkar replied:
Moved this function to misc.c

bhroam commented on src/scheduler/fifo.c:
```diff
@@ -2355,6 +2351,46 @@ next_job(status *policy, server_info *sinfo, int flag)
+
+	if (src !=NULL) {
+		len = strlen(src);
+		*dest = (char*)malloc(len + 1);
```
The scheduler style doesn't cast the output of malloc().  It returns a void* pointer which can be assigned into any pointer.

suresh-thelkar replied:
Done

bhroam replied:
Is this coming in a future commit?  I don't see it done.

suresh-thelkar replied:
yes

bhroam commented on src/scheduler/fifo.c:
```diff
@@ -2355,6 +2351,46 @@ next_job(status *policy, server_info *sinfo, int flag)
+		len = strlen(src);
+		*dest = (char*)malloc(len + 1);
+		if (*dest == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
```
Don't use log_err(), use schdlog().

suresh-thelkar replied:
Most of the places where the macro MEM_ERR_MSG is used, it seems we are using log_err() only.  I am fine to do the change but eager to know the exact difference between these two in terms of usage. As far as I know schdlog() logs to both the scheduler logs as well as console. Is this something that we actually want in this case?

bhroam replied:
The scheduler doesn't use log_events like the server does.  It has its own thing (log_filter).  The schdlog() function filters based on it.  If you use log_err(), then there is no way to filter these messages if it is desired.

suresh-thelkar replied:
Thanks for letting me know this. I have made the changes accordingly. It seems there are several places where still log_err() is used with MEM_ERR_MSG which lead to my initial confusion.

bhroam commented on src/scheduler/pbs_sched.c:
```diff
@@ -791,7 +792,7 @@ are_we_primary()
 	}
 	strncpy(scheduler_host_name, server_host, sizeof(scheduler_host_name));
-	scheduler_host_name [ sizeof(scheduler_host_name) -1 ] = '\0';
+	scheduler_host_name[sizeof(scheduler_host_name) -1] = '\0';
```
Add a space after '-'

suresh-thelkar replied:
Done

bhroam commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -765,7 +766,7 @@ are_we_primary()
 	}
 	strncpy(scheduler_host_name, server_host, sizeof(scheduler_host_name));
-	scheduler_host_name [ sizeof(scheduler_host_name) -1 ] = '\0';
+	scheduler_host_name[sizeof(scheduler_host_name) -1] = '\0';
```
add a space after '-'

suresh-thelkar replied:
Done

bhroam commented on src/server/req_select.c:
```diff
@@ -365,7 +366,8 @@ req_selectjobs(struct batch_request *preq)
 	 */
-	if ((find_sched_from_sock(preq->rq_conn) != NULL) && (!scheduler_jobs_stat)) {
+	psched = find_sched_from_sock(preq->rq_conn);
+	if ((psched != NULL) && (psched == dflt_scheduler ) && (!scheduler_jobs_stat)) {
```
There's an extra space after dflt_scheduler.

suresh-thelkar replied:
Done

bhroam commented on src/include/pbs_sched.h:
```diff
@@ -41,13 +41,23 @@
 #include "sched_cmds.h"
-#include "net_connect.h"
+#include "work_task.h"
+#include  "net_connect.h"
```
There's an extra space after include

suresh-thelkar replied:
Done

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -2990,9 +3000,9 @@ try_connect_database(pbs_db_conn_t *conn)
 		db_oper_failed_times++;
 		get_db_errmsg(failcode, &db_err_msg);
-		log_set_dberr(db_err_msg, conn->conn_db_err);
+			log_set_dberr(db_err_msg, conn->conn_db_err);
```
indentation is off here

suresh-thelkar replied:
Done

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -3109,7 +3119,7 @@ setup_db_connection(char *host, int timeout, int have_db_control)
 		if (db_err_msg && strlen(db_err_msg) > 0) {
 			log_event(PBSEVENT_SYSTEM | PBSEVENT_FORCE, PBS_EVENTCLASS_SERVER, LOG_CRIT, msg_daemonname, db_err_msg);
-			fprintf(stderr, "%s\n", db_err_msg);
+		fprintf(stderr, "%s\n", db_err_msg);
```
Indentation is off here

suresh-thelkar replied:
Done

arungrover commented on src/server/req_jobobit.c:
```diff
@@ -649,7 +651,14 @@ rel_resc(job *pjob)
+	else {
+		pbs_queue *pq;
+		pq = find_queuebyname( pjob->ji_qs.ji_queue);
+		sprintf(log_buffer, "Unable to reach scheduler associated with job %s", pq->qu_attr[QA_ATR_partition].at_val.at_str);
```
Design document states that the message gets printed with a job id not partition id.

suresh-thelkar replied:
Have updated the design document.

arungrover replied:
well, the message still needs to change to say "partition" instead of "job"

arungrover commented on src/server/req_runjob.c:
```diff
@@ -331,15 +333,22 @@ req_runjob(struct batch_request *preq)
+	if (!find_assoc_sched_jid(jid, &psched)) {
+		sprintf(log_buffer, "Unable to reach scheduler associated with job %s", jid);
+		log_err(-1, __func__, log_buffer);
+		req_reject(PBSE_IVALREQ, 0, preq);
```
I think we should reject the request with a new error number which says corresponding scheduler not found

arungrover commented on src/server/run_sched.c:
```diff
@@ -155,6 +159,132 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+ * 	   - 0 if fail
+ */
+int
+find_assoc_sched_jid(char *jid, pbs_sched **target_sched)
```
why don't we expect a job structure to be passed to the function. It will probably save you a find job call.
I am getting a very bleak memory that I have asked this before too :)

suresh-thelkar replied:
Not all the places where we call this function job structure is not available in which case we any way have to do find_job() call

arungrover commented on src/server/run_sched.c:
```diff
@@ -155,6 +159,132 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+	if (pj == NULL)
+		return 0;
+
+	q_name = pj->ji_qs.ji_queue;
```
at this point why don't you call find_assoc_sched_pque? code is almost same in both the functions

suresh-thelkar replied:
Implemented.

arungrover commented on src/server/run_sched.c:
```diff
@@ -155,6 +159,132 @@ put_sched_cmd(int sock, int cmd, char *jobid)
 
+/**
+ * @brief
+ * 		find_assoc_sched - find the corresponding scheduler which is responsible
```
please update the function header

suresh-thelkar replied:
Implemented.

arungrover commented on src/server/run_sched.c:
```diff
@@ -155,6 +159,132 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+
+/**
+ * @brief
+ * 		find_assoc_sched - find the corresponding scheduler which is responsible
```
please update the function header

suresh-thelkar replied:
Done

arungrover commented on src/server/run_sched.c:
```diff
@@ -155,6 +159,132 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+ * 		find_assoc_sched - find the corresponding scheduler which is responsible
+ * 		for handling this job.
+ *
+ * @param[in]	jid	- job id
```
this would be queue

suresh-thelkar replied:
Done

suresh-thelkar commented on src/lib/Liblog/pbs_messages.c:
```diff
@@ -405,6 +405,8 @@ char *msg_cannot_set_route_que = "Route queues are incompatible with the partiti
 char *msg_partition_not_in_queue = "Partition %s is not part of queue for node";
 char *msg_invalid_partion_in_queue = "Invalid partition in queue";
+char *msg_sched_op_not_permitted = "Operation is not permitted on default scheduler";
+char *msg_sched_part_already_used = "Partition is already associated with other scheduler";
```
Getting scheduler name is trivial but getting partition is not straightforward since it logically groups nodes/queues and hence planning to take this as part of subsequent tickets. Also I have added some replies to the same comment in the previous PR. Please look into that.

bhroam commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,600 @@
+        self.server.manager(MGR_CMD_SET, SCHED, a, id="sc5")
+        sched = self.server.schedulers['sc5']
+        # Try starting without sched_priv and sched_logs
+        ret = self.server.schedulers['sc5'].start(sched_port='15055')
```
I was unaware that PTL already had a self.schedulers.  From what I can see self.schedulers is used when -p servers is used.  It needs to be modified anyway because right now it will only be a list of default schedulers.  There won't be any way to get at the non-default schedulers on the other servers.  I suggest that you make self.schedulers a list of dictionaries.  The list will be each of the -p servers, and the dictionary will be a dictionary of schedulers.  So self.schedulers[0]['default'] is self.scheduler.  It's only used in two places in our tests, so it won't be hard to update.

arungrover replied:
I agree with Bhroam, you could just make it a list of dictionaries and it will resolve the problem.

visheshh replied:
Hi @bhroam , @arungrover , So the dictionary of self.schedulers will contain the hostname provided in -p and list of schedulers per host .
But do you think we should call all scheduler functions using self.schedulers[0]['sched_name'] ?
Or is it fine to call self.server.schedulers['sched_name'].func_name(), We are currently using self.server.schedulers dictionary values such log_path and priv_path inside Class Sched functions .

bhroam replied:
I dislike self.server.schedulers because a scheduler is a daemon and not part of the server.  It isn't like a queue which is a server concept.  It is true that the scheduler information is held at the server, but so is mom information held by the vnodes.  We should be using self.schedulers.

arungrover commented on src/server/sched_func.c:
```diff
@@ -331,13 +381,28 @@ int
+		if (actmode == ATR_ACTION_ALTER) {
+			if (pattr->at_val.at_long)
+				set_scheduler_flag(SCH_SCHEDULE_CMD, (pbs_sched *)pobj);
+	}
```
Indentation seems to be off here

suresh-thelkar replied:
Done

arungrover commented on src/server/sched_func.c:
```diff
@@ -331,13 +381,28 @@ int
 	} else {
-		/*TODO need add the code for other scheduler apart from main scheduler. */
+		if (actmode == ATR_ACTION_ALTER) {
+			if (pattr->at_val.at_long)
```
Why do we check for a value? Does this mean that one can not set the scheduling attribute on a multisched to 0?

suresh-thelkar replied:
We only need to set the flag SCH_SCHEDULE_CMD if scheduling is set to true. To answer your question yes one can set scheduling of multisched to false if they want to turn off scheduling it works fine.

arungrover commented on src/server/sched_func.c:
```diff
@@ -331,13 +381,28 @@ int
+				(void)sched_save_db(dflt_scheduler, SVR_SAVE_FULL);
+			}
+		} else {
+			server.sv_attr[SRV_ATR_scheduling].at_val.at_long = pattr->at_val.at_long;
```
when do we reach this else case? If we come here when we are trying to set scheduling on a scheduler object then why do we set server's attribute here? I'm confused.

suresh-thelkar replied:
The else block is hit when admin tries to set scheduling on default scheduler. In this case we also have to make sure that server's scheduling attribute is set to the same value as that of default scheduler.  This is to make sure that data is consistent between default scheduler vs server. Also making sure that vice versa case is also consistent.

visheshh commented:
Hi @bhroam , I have addressed most of the review comments .
Killing of schedulers in cleanup and self.schedulers comments are still not addressed .
I will address in next commit with Varun's comments . Please review the requested changes.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ */
+static int
+copy_attr_value(char **dest, char *src)
```
We can not actually make use of pbs_strcat() for this scenario it is because this function expects at least dest is pre-allocated with some memory prior to this call(and also we need to pass the initial size of dest to this function) which is not true in our case. Also as per Bhroam's earlier comment I have already moved this function to misc.c

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ */
+static int
+sched_settings_frm_svr(struct batch_status *status)
```
We are actually taking care of using the same values set by admin through qmgr. But after scheduler validating these attributes and in case if it finds that they do not pass validation criteria then we are just setting the corresponding scheduler's comment to the reason why the corresponding validation failed  and we are not changing any of the attributes that he has set. I think this is certainly needed which actually gives an indication to the admin that he has to correct the faulty values set for the sched object by him.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+			} else if (!strcmp(attr->name, ATTR_partition)) {
+				if (copy_attr_value(&tmp_partitions, attr->value) == 0)
+					return 0;
+			} else if (!strcmp(attr->name, ATTR_comment)) {
```
By getting the comment set for a sched object back from server we can decide whether to clear the comment or not in case if at this point of time if the admin sets-up right value for the attribute for which the comment is set earlier. If we are not doing this we have to always clear the comment no matter it is set or not on the sched object each time when the user changes the sched object values correctly.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+		int priv_dir_update_fail = 0;
+		struct attropl *patt;
+		char comment[MAX_LOG_SIZE] = {0};
+		static char *priv_dir = NULL;
```
Implemented

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+					log_err(errno, __func__, log_buffer);
+					free(tmp_log_dir);
+					free(tmp_priv_dir);
+					free(tmp_partitions);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+					log_err(errno, __func__, log_buffer);
+					free(tmp_log_dir);
+					free(tmp_priv_dir);
+					free(tmp_partitions);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+					free(tmp_partitions);
+				}
+
+				strncpy(comment, "Unable to change the sched_logs directory", MAX_LOG_SIZE -1);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+				patt->next = patt + 1;
+				patt++;
+				patt->name = ATTR_scheduling;
+				patt->value = "0";
```
This is done as per the suggestion of Bhroam. This will ensure that jobs will not be scheduled by a corresponding scheduler if its configuration set by the admin is invalid.

bhroam replied:
I guess we have two options here.  We could set scheduling to false or we can just mark the scheduler down.  I'm fine with either.

bhroam replied:
Now that I think about it more, I like setting scheduling to false as well.  Having a scheduler as down has a specific meaning.  It means the server can not connect to the scheduler.  It would muddy the waters if we had it mean multiple things.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+				snprintf(log_buffer, sizeof(log_buffer), "scheduler log directory is changed to %s", tmp_log_dir);
+				schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_SCHED, LOG_INFO,
+						"reconfigure", log_buffer);
+				free(log_dir);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+		log_dir = tmp_log_dir;
+
+		if ((priv_dir != NULL) && strcmp(priv_dir, tmp_priv_dir) != 0) {
+			int c;
```
I think it is not needed to initialize the variable 'c' here as the function chk_file_sec() makes sure that it will return the appropriate value.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+					priv_dir_update_fail = 1;
+				} else {
+					int lockfds;
+					(void)unlink("sched.lock");
```
Two schedulers can not have the same priv directory. This is not possible since we bail out at the early stage itself even if the admin tries to give the same directory for multiple sched objects.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+				strncpy(comment, "Unable to change the sched_priv directory", MAX_LOG_SIZE);
+				free(tmp_log_dir);
+				free(tmp_priv_dir);
+				free(tmp_partitions);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,357 @@ next_job(status *policy, server_info *sinfo, int flag)
+			}
+			return 0;
+		}
+		if (cstrcmp(partitions, tmp_partitions) != 0) {
```
No type here.

arungrover replied:
ahh...I get it, it is our internal function :) how did we come up with such name :)

bhroam replied:
I wrote it a very long time ago.  I guess I meant check strcmp().  I just got tired of writing the code to make sure both arguments were not NULL before calling strcmp().

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -128,6 +129,7 @@ def test_permission_on_scheduler(self):
                             SCHED,
                             id="testDeleteSched",
                             runas=ROOT_USER)
+        print self.server.schedulers
```
This looks like a debug statement left in.  If it isn't, use self.logger() instead of print.

bhroam commented:
Codacy is complaining about an unused import in the tests.

suresh-thelkar commented on src/scheduler/node_info.c:
```diff
@@ -207,8 +209,10 @@ query_nodes(int pbs_sd, server_info *sinfo)
 	}
 #endif /* localmod 049 */
 
+	my_partitions = break_comma_list(partitions);
```
This statement is not part of this function now, instead it is moved to the function node_in_partition() or queue_in_partition()  as suggested by Bhroam and the appropriate action is taken to free my_partitions.

suresh-thelkar commented on src/scheduler/node_info.c:
```diff
@@ -309,6 +339,13 @@ query_node_info(struct batch_status *node, server_info *sinfo)
+		else if(!strcmp(attrp->name, ATTR_partition)) {
+			ninfo->partition = string_dup(attrp->value);
+			if (ninfo->partition == NULL) {
+				log_err(errno, "node_queue_info", MEM_ERR_MSG);
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -143,19 +148,13 @@ char		*configfile = NULL;	/* name of file containing
 char		**glob_argv;
 char		usage[] =
-	"[-d home][-L logfile][-p file][-S port][-R port][-n][-c clientsfile]";
+	"[-d home][-L logfile][-p file][-I schedname][-S port][-R port][-n][-c clientsfile]";
```
I have added this to design document with an example.

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -1307,16 +1257,18 @@ main(int argc, char *argv[])
+		saddr.sin_port = htons(sched_port);
+		saddr.sin_addr.s_addr = INADDR_ANY;
+		if (bind(server_sock, (struct sockaddr *)&saddr, sizeof(saddr)) < 0) {
+			errno = WSAGetLastError();
```
Added back the directive.

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -1307,16 +1257,18 @@ main(int argc, char *argv[])
+		if (bind(server_sock, (struct sockaddr *)&saddr, sizeof(saddr)) < 0) {
+			errno = WSAGetLastError();
+			log_err(errno, __func__, "bind");
+			die(0);
```
Yes the scheduler won't come up if the port given is busy. Admin has to change its port through qmgr and then start it again.

suresh-thelkar commented on src/scheduler/queue_info.c:
```diff
@@ -381,6 +411,13 @@ query_queue_info(status *policy, struct batch_status *queue, server_info *sinfo)
+		else if(!strcmp(attrp->name, ATTR_partition)) {
+			qinfo->partition = string_dup(attrp->value);
+			if (qinfo->partition == NULL) {
+				log_err(errno, "query_queue_info", MEM_ERR_MSG);
```
Done

suresh-thelkar commented on src/server/node_manager.c:
```diff
@@ -3032,7 +3035,12 @@ deallocate_job(mominfo_t *pmom, job *pjob)
+	if (find_assoc_sched_jid(pjob->ji_qs.ji_jobid, &psched))
+		set_scheduler_flag(SCH_SCHEDULE_TERM, psched);
+	else {
+		sprintf(log_buffer, "Unable to reach scheduler associated with partition");
```
Done

suresh-thelkar commented on src/server/node_manager.c:
```diff
@@ -8685,7 +8694,12 @@ free_sister_vnodes(job *pjob, char *vnodelist, char *err_msg,
+	if (find_assoc_sched_jid(pjob->ji_qs.ji_jobid, &psched))
+		set_scheduler_flag(SCH_SCHEDULE_TERM, psched);
+	else {
+		sprintf(log_buffer, "Unable to reach scheduler associated with partition");
```
Done

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -327,30 +315,41 @@ def test_job_queue_partition(self):
-        if the partitions are set on the scheduler
+        if the partitions are set on the scheduler and test that even a job
+        submitted to partition P1 queue can run on node set to partition P4
+        if both partitions are serviced by same scheduler.
```
Hi @bhroam ,
Are you saying the comment does not sound right or the current scheduler behaviour ?
If the P1 and P4 and serviced by different schedulers , Then a job submitted to queue1 which is set to P1 will never run on node set to P4 . But now it can run . Because nodes are not assigned to queues , partitions are assigned . So scheduler can run all its available jobs on all the nodes available to it.

bhroam replied:
I'm saying the behavior is wrong.  Job's should be contained in their partitions.  They shouldn't escape them just because the scheduler servicing the partition also services other partitions.  Think of the case where each partition is part of a machine owned by a particular department.  The department won't like it when jobs from another department suddenly starts running on their hardware.

suresh-thelkar commented on src/server/svr_func.c:
```diff
@@ -801,9 +800,9 @@ set_rpp_highwater(attribute *pattr, void *pobj, int actmode)
  */
 void
-set_sched_sock(int s)
+set_sched_sock(int s, pbs_sched *psched)
```
Done

suresh-thelkar commented:
After pushing the latest commits, I observed conflicts. I am rebasing and push the changes after testing shortly.

bhroam commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -544,11 +550,28 @@ def test_job_sorted_per_scheduler(self):
+    def test_qrun_job(self):
+        """
+        Test jobs are sorted as per job_sort_formula
+        inside each scheduler
```
How does this test the job_sort_formula?  There is only one job.

visheshh replied:
I will change the comment , It is test of qrun.

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -55,28 +53,7 @@ def setUp(self):
-                             os.path.join(sched_home, sched_logs_dir),
-                             recursive=True)
-        self.server.schedulers[sched_name].start(sched_port, sched_home)
+        self.sched_configure("TestCommonSched")
```
You're calling sched_configure(), but you just removed the function.

visheshh replied:
I have added it in pbs_testsuite.py as per varun's suggestion as i am using it in both functional and interface test suites .

bhroam replied:
I disagree with @varunsonkar.  While duplicating code is bad, putting test specific functions in the parent class to all tests is worse.  The parent class should remain generic to all tests.

visheshh replied:
Hi @bhroam , 
To start all the schedulers even in future we need to create a priv directory and log directory , Before starting the scheduler . So i had written this function . It can used in future in all the test suites which might need one or more scheduler . Wont keeping it in pbs_testsuite.py help ?

bhroam replied:
This function does sound useful.  I think you should modify it some and put it as part of the scheduler object in PTL itself.  Maybe instead of sched_configure(), you write create_scheduler().  You pass in the name of the scheduler and create a working scheduler.  I'd also not do a copy of the sched_log directory.  Just make the directory.  You don't want to start matching logs from the default scheduler.

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -269,80 +271,78 @@ def test_remove_partition_sched(self):
-        # self.server.manager(MGR_CMD_UNSET, SCHED, 'partition',
-        #                    id="sc2", expect=True)
+        self.server.manager(MGR_CMD_UNSET, SCHED, 'partition',
+                            id="sc2", expect=True)
```
I am testing that unset sched partition removes partition attribute from qmgr.

visheshh replied:
removed as currently removal of any attribute cannot be tested as it is blocked PP-1202

visheshh commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -999,6 +1007,29 @@ def stop_proc_monitor(self):
         self.metrics_data['procs'] = self._procmon.db_proc_info
         self._process_monitoring = False
 
+    def sched_configure(self, sched_name, sched_home=None):
```
will it be fine if i change it to sched_start .

visheshh commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -999,6 +1007,29 @@ def stop_proc_monitor(self):
+        if not os.path.exists(os.path.join(sched_home, sched_logs_dir)):
+            self.du.mkdir(path=os.path.join(sched_home, sched_logs_dir),
+                          sudo=True)
+        self.server.schedulers[sched_name].start(sched_home)
```
I have self.scheds , But the dict key is the hostname or the server name and then the values is again a dictionary of scheduler names and its objects.
So it has to something like self.schedulers[hostname][sched_name] . If this is okay i will change it.

bhroam replied:
The dict of self.schedulers is consistent with self.moms.  It's a dictionary whose keys are hostnames.  It would make sense that we make it a dictionary of scheduler names as well.  Since most tests will only use schedulers associated with one server, it might make sense to also create a shorthand for one server.  This way you can do something like self.scheds['name'] and it means self.schedulers[self.mom.shortname]['name']

visheshh replied:
I was not using self.scheds for anything other than intializing and iterating through hostnames .
Now i have created a shorthand for self.server.schedulers and will update all the test cases.

suresh-thelkar commented on src/server/svr_func.c:
```diff
@@ -927,6 +926,31 @@ ssignon_transition_okay(attribute *pattr, void *pobject, int actmode)
+action_svr_iteration(attribute *pattr, void *pobj, int mode)
+{
+	/* set this attribute on main scheduler */
+	if (dflt_scheduler) {
```
Made the code changes to add only the required modes.

suresh-thelkar commented:
@bhroam / @arungrover 
I have added 5 commits to the pull request which addresses all of your comments given earlier along with the additional code changes for couple of requirements that are needed.

Out of these 5 commits please ignore e13c4cf for the time being. I had to do this last week as we were getting some intermittent issues with PTL which can not be easily traceable. @visheshh  is currently working on finding the root cause and fixing this and we can put back the code changes required for this commit as early as possible. In the interest of time to proceed further to get your valuable comments on other things that we have worked we reverted this temporally.

Please review the latest commits and give your comments if any at the earliest.

bhroam commented on src/scheduler/check.c:
```diff
@@ -1607,22 +1611,39 @@ check_nodes(status *policy, server_info *sinfo, queue_info *qinfo, resource_resv
+			ninfo_arr_filtrd[idx++] = ninfo_arr[i];
+		}
+	}
+	ninfo_arr_filtrd[idx] = NULL;
```
Please do this work once and store it on the queue and not once per job.  Doing it once per job will significantly hurt performance.

suresh-thelkar replied:
As you suggested now doing the filtering only once at the start of the scheduling cycle.

bhroam commented on src/server/req_manager.c:
```diff
@@ -1625,9 +1625,13 @@ mgr_server_unset(struct batch_request *preq)
-				dflt_scheduler->sch_attr[SCHED_ATR_schediteration].at_flags |=
-						ATR_VFLAG_SET | ATR_VFLAG_MODIFY | ATR_VFLAG_MODCACHE;
+				svrattrl *tm_list;
+				tm_list = attrlist_create(plist->al_name, NULL, 0);
```
check tm_list for NULL

suresh-thelkar replied:
Implemented.

bhroam commented on src/server/req_manager.c:
```diff
@@ -1683,10 +1687,12 @@ mgr_sched_set(struct batch_request *preq)
 		reply_badattr(rc, bad_attr, plist, preq);
 	else {
-
+		//int unset_flag = 0;
```
Either uncomment this or remove it.

suresh-thelkar replied:
Done

bhroam commented on src/server/req_manager.c:
```diff
@@ -1683,10 +1687,12 @@ mgr_sched_set(struct batch_request *preq)
 		(void)sched_save_db(psched, SVR_SAVE_FULL);
-		set_sched_default(psched);
+		/*if (preq->rq_fromsvr)
+			unset_flag = 1;*/
```
Either uncomment this or remove it.

bhroam commented on src/server/req_manager.c:
```diff
@@ -1723,6 +1729,17 @@ mgr_sched_unset(struct batch_request *preq)
+			if (dflt_scheduler) {
+				svrattrl *t_list;
+				t_list = attrlist_create(tmp_plist->al_name, NULL, 0);
+				t_list->al_link.ll_next->ll_struct = NULL;
```
Check t_list for NULL

suresh-thelkar replied:
Implemented

bhroam commented on src/server/sched_func.c:
```diff
@@ -412,10 +414,13 @@ poke_scheduler(attribute *pattr, void *pobj, int actmode)
+  */
 void
-set_sched_default(pbs_sched* psched)
+set_sched_default(pbs_sched* psched, int unset_flag)
```
Attach the * to the variable name

suresh-thelkar replied:
Done

bhroam commented on src/server/svr_func.c:
```diff
@@ -798,7 +798,8 @@ set_rpp_highwater(attribute *pattr, void *pobj, int actmode)
  *		state attribute cache.
  *
- * @param[in]	s	-	internal socket used to communicate with the scheduler
+ * @param[in] s		-	internal socket used to communicate with the scheduler
```
Is the indentation off?  It shows off on GitHub

suresh-thelkar replied:
It looks like it is GitHub issue. The indentation seems fine in my editor.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -2183,7 +2183,7 @@ resresv_set_use_queue(queue_info *qinfo)
 	if (qinfo->has_hard_limit || qinfo->has_soft_limit || qinfo->has_nodes ||
 	    qinfo->is_ded_queue || qinfo->is_prime_queue || qinfo->is_nonprime_queue ||
-	    qinfo->has_resav_limit || qinfo->resv != NULL)
+	    qinfo->has_resav_limit || qinfo->resv != NULL || (qinfo->partition != NULL))
```
This will cause any queue with a partition set to get its own equivalence class.  This isn't quite right.  If multiple queues share the same partition, you'll create multiple equivalence classes where there should only be one.  The right way to do this is to add partition to resresv_set.  If it is NULL, it'll be ignored.  If it is not NULL, it'll be used to break jobs into sets.

suresh-thelkar replied:
Bhroam as discussed I have added the changes for equivalences classes as suggested. I have also done unit testing and all the test cases are passing.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -3243,6 +3243,18 @@ is_vnode_eligible(node_info *node, resource_resv *resresv,
+				return 0;
+			}
+		}
+	}
```
Doing string comparisons in the scheduler is bad.  Doing them in node search is horrible.  You'll significantly hurt scheduler performance.  Create the list of nodes associated to a queue's partition on the queue itself.  This means we only are searching the right set of nodes.

suresh-thelkar replied:
Implemented.

arungrover commented on src/server/node_manager.c:
```diff
@@ -1949,9 +1950,9 @@ find_degraded_occurrence(resc_resv *presv, struct pbsnode *np,
-				occr_time = get_occurrence(rrule, dtstart, tz, j);
+		/* we keep track of the occurrence time to determine the earliest
+		 * degraded time */
+		occr_time = get_occurrence(rrule, dtstart, tz, j);
```
Indentation is off here too

suresh-thelkar replied:
Done

arungrover commented on src/server/node_manager.c:
```diff
@@ -1949,9 +1950,9 @@ find_degraded_occurrence(resc_resv *presv, struct pbsnode *np,
-				/* we keep track of the occurrence time to determine the earliest
-				 * degraded time */
-				occr_time = get_occurrence(rrule, dtstart, tz, j);
+		/* we keep track of the occurrence time to determine the earliest
```
Please indent this comment and the line below.

suresh-thelkar replied:
Done

arungrover commented on src/server/node_manager.c:
```diff
@@ -3033,7 +3036,12 @@ deallocate_job(mominfo_t *pmom, job *pjob)
+		set_scheduler_flag(SCH_SCHEDULE_TERM, psched);
+	else {
+		sprintf(log_buffer, "Unable to find scheduler associated with partition");
+		log_err(-1, __func__, log_buffer);
```
I guess you can just give the string here. copying message into buffer isn't helping

suresh-thelkar replied:
Done

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -370,8 +366,8 @@ int tpp_network_up = 0;
-    tpp_network_up = 1;
+	log_tppmsg(LOG_INFO, NULL, "net restore handler called");
+	tpp_network_up = 1;
     ping_nodes(NULL);
```
Please indent this line too

suresh-thelkar replied:
Done

arungrover commented on src/server/req_delete.c:
```diff
@@ -327,6 +329,7 @@ req_deletejob(struct batch_request *preq)
 	int delhist = 0;
 	int maxindex = 0;
 	int count = 0;
+	extern job  *chk_job_request(char *, struct batch_request *, int *);
```
why is this line needed here

suresh-thelkar replied:
Thanks for catching this. This is now not required since all the compiler warnings are resolved.

suresh-thelkar commented on src/scheduler/misc.c:
```diff
@@ -1701,3 +1701,44 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+	int ret = 0;
+
+	if (*dest != NULL)
+		free(*dest);
```
@arungrover 
Instead of using free I am now using realloc() as per the earlier comment from Bhroam. So I don't think the problem you mentioned will arise now.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2328,3 +2351,329 @@ next_job(status *policy, server_info *sinfo, int flag)
+		err = pbs_manager(connector,
+				MGR_CMD_UNSET, MGR_OBJ_SCHED,
+			sc_name, attribs, NULL);
+		free(attribs);
```
Done

suresh-thelkar commented on src/scheduler/misc.c:
```diff
@@ -1701,3 +1701,44 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+ *	Helper function used to copy the given source string to the destination string. It also
+ *	frees the destination and then allocates required memory before copying.
+ *
+ * @param[in] dest - Address of pointer to destination string
```
Done. This is now both in and out parameter as per the current code changes.

suresh-thelkar commented on src/scheduler/misc.c:
```diff
@@ -1701,3 +1701,44 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+ *	frees the destination and then allocates required memory before copying.
+ *
+ * @param[in] dest - Address of pointer to destination string
+ * @param[in] dest - pointer to source string
```
Corrected the typo.

suresh-thelkar commented on src/scheduler/misc.c:
```diff
@@ -1701,3 +1701,44 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+			schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, MEM_ERR_MSG);
+			return ret;
+		}
+		strncpy(*dest, src, len);
```
Done

suresh-thelkar commented on src/scheduler/node_info.c:
```diff
@@ -1270,6 +1283,14 @@ dup_node_info(node_info *onode, server_info *nsinfo,
 	nnode->NASrank = onode->NASrank;
 #endif /* localmod 049 */
 
+	if (nnode->partition != NULL) {
```
Done

suresh-thelkar commented on src/scheduler/node_info.c:
```diff
@@ -5639,3 +5672,44 @@ check_node_array_eligibility(node_info **ninfo_arr, resource_resv *resresv, plac
+ *
+ * @param[in]	ninfo		-  node information
+ *
+ * @return	a node_info filled with information from node
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -1521,14 +1475,14 @@ main(int argc, char *argv[])
-				rc = set_tpp_config(&pbs_conf, &tpp_conf, nodename, pbs_conf.scheduler_service_port,
-												pbs_conf.pbs_leaf_routers, pbs_conf.pbs_use_compression,
-												TPP_AUTH_RESV_PORT, NULL, NULL);
+		rc = set_tpp_config(&pbs_conf, &tpp_conf, nodename, sched_port,
```
Done

suresh-thelkar commented on src/scheduler/queue_info.c:
```diff
@@ -1000,3 +1022,44 @@ node_queue_cmp(node_info *ninfo, void *arg)
+ *
+ * @param[in]	qinfo		-  queue information
+ *
+ * @return	a node_info filled with information from node
```
Done

suresh-thelkar commented on src/server/node_manager.c:
```diff
@@ -176,6 +176,7 @@
 #include	"hook_func.h"
 #include	"sched_cmds.h"
 #include	"provision.h"
+#include "pbs_sched.h"
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2456,14 +2416,16 @@ sched_settings_frm_svr(struct batch_status *status)
 				}
 
-				strncpy(comment, "Unable to change the sched_logs directory", MAX_LOG_SIZE -1);
+				strncpy(comment, "Unable to change the sched_log directory", MAX_LOG_SIZE -1);
```
Done

suresh-thelkar commented on src/scheduler/misc.c:
```diff
@@ -1701,3 +1701,44 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+ *
+ */
+int
+copy_attr_value(char **dest, char *src)
```
I have changed the name of the function to strcpy_alloc. Hope you like the name. If not please suggest a better name.

suresh-thelkar commented on src/scheduler/misc.c:
```diff
@@ -1701,3 +1701,44 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+	if (*dest != NULL)
+		free(*dest);
+
+	if (src !=NULL) {
```
Done

suresh-thelkar commented on src/scheduler/misc.c:
```diff
@@ -1701,3 +1701,44 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+	if (src !=NULL) {
+		int len = 0;
+		len = strlen(src);
+		*dest = (char*)malloc(len + 1);
```
Done. As per your earlier comment I am now using realloc and also avoiding the typecast.

suresh-thelkar commented on src/scheduler/node_info.c:
```diff
@@ -5689,3 +5660,44 @@ check_node_array_eligibility(node_info **ninfo_arr, resource_resv *resresv, plac
+		log_err(errno, __func__, "Error parsing partitions");
+		return 0;
+	}
+	for (i=0; my_partitions[i] != NULL; i++) {
```
Implemented.

suresh-thelkar commented on src/scheduler/queue_info.c:
```diff
@@ -1048,3 +1022,44 @@ node_queue_cmp(node_info *ninfo, void *arg)
+		log_err(errno, __func__, "Error parsing partitions");
+		return 0;
+	}
+	for (i=0; my_partitions[i] != NULL; i++) {
```
Done

suresh-thelkar commented:
@bhroam / @arungrover 

I have pushed the following commits where in I have implemented all of your comments. Please review it and give your further review comments if any or please provide your sign off.

b424b40 Adding documentation changes for log messages as suggsted by Anne
b7aa380 Implementing Arun's review comments
586ae96 Fixing the preemption issue for Multisched...
dc7c654 Implementing Bhroam's latest review comments

arungrover commented:
@suresh-thelkar I could finally get the pull request to open. I'm afraid that if I add more comments then it will only make matter worse.
I'm closing it and request you to raise another pull request.
